File: 2021/day1/day1.py
def part1(depths: list[int]):
    nr_increases = 0

    for n in range(1, len(depths)):
        if depths[n] > depths[n - 1]:
            nr_increases += 1

    print('Part 1:', nr_increases)


def part2(depths: list[int]):
    nr_increases = 0
    prev_sum = 0

    for n in range(len(depths) - 2):
        if prev_sum == 0:
            prev_sum = depths[n] + depths[n + 1] + depths[n + 2]
            continue

        current_sum = depths[n] + depths[n + 1] + depths[n + 2]

        if current_sum > prev_sum:
            nr_increases += 1

        prev_sum = current_sum

    print('Part 2:', nr_increases)


if __name__ == '__main__':
    with open('in.txt') as file:
        lines = file.read().splitlines()

    depths = [int(line) for line in lines]

    part1(depths)
    part2(depths)

File: 2021/day1/day1.hs
module Main where

import System.IO

main :: IO ()
main = do
  contents <- readFile "in.txt"
  let depths = map (\line -> read line :: Int) $ lines contents
  putStrLn $ "Part 1: " ++ show (part1 depths)
  putStrLn $ "Part 2: " ++ show (part2 depths)

part1 :: [Int] -> Int
part1 depths = foldl (\acc i -> if (depths !! i) > (depths !! (i - 1)) then acc + 1 else acc) 0 [1 .. length depths - 1]

part2 :: [Int] -> Int
part2 depths = foldl (\acc i -> if sum3 depths i > sum3 depths (i - 1) then acc + 1 else acc) 0 [3 .. length depths - 1]

sum3 :: [Int] -> Int -> Int
sum3 depths i = depths !! i + depths !! (i - 1) + depths !! (i - 2)
